### PR TITLE
feat: implement the dht for the market server

### PR DESCRIPTION
**Pull request by reaovyd**
# Orcanet Market

This is our team (Seahorse)'s implementation for the Orcanet Market server; it simply just serves as a PoC and somewhat of a minimum viable product (if you don't count dealing with remote nodes). I haven't created the README.md(s) yet (but will do so soon) so some information will be here, which would be included by a future README.md(s).

## Crates
- `market_dht` - This is the core library that provides the DHT for a peer node. It is modular in the sense that this can be embedded inside a peer node implementation itself or, of course, inside the gRPC `market_server`. It is based off the `tokio` async executor (for now?? maybe agnostic in the future?)
- `market_server` - (might bundle as an example under `market_dht/examples/grpc_server`) provides a gRPC server based on the [market protofiles](https://github.com/sbu-416-24sp/orcanet-market-go). This gRPC server accepts a peer port (perhaps in the future, if needed, will also provide an external address as a gateway for entrance into the machine where the peer lies) and incorporates the [`market_dht::peer::Peer`](https://github.com/reaovyd/orcanet-market-rust/blob/main/market_dht/src/peer.rs) as the way to communicate with the DHT through provided methods that are equivalent to the gRPC methods.
- `market_test_client` - (might bundle as an example under `market_dht/examples/grpc_server`) provides a simple test CLI client for communicating with the market gRPC server found in `market_server`. 
- `market_proto` - (might bundle as an example under `market_dht/examples/grpc_server`) provides and generates the gRPC functionality based on the [`proto_files`](https://github.com/sbu-416-24sp/orcanet-market-go)

### `market_dht`
As mentioned above, this is the core library that handles the operations from DHT peer clients. We have used the `identify` protocol for peer discovery and, of course, the `Kademlia` protocol for peer routing. We, also, further used the `request_response` protocol for handling a custom message for handling how we can get supplier information (to be explained later). These are the minimum number of protocols that we used for this MVP. We chose not to use mDNS, for now, to explore the behavior of how things would typically work with what would appear to be a remote node (of course, after the NAT traversal, I think?), but we may include mDNS in the future.

#### Logs
There is log functionality in here using the abstract `log` crate. We haven't configured tracing instruments for where we are logging yet, but will do so in the future and possibly just move everything to a `DEBUG` level since there are many logs being printed that were just useful for us (though does not break user functionality).

#### Provider Record TTL and Republishing
Each provider record has a TTL of 1 hour as defined in the system design document. Each provider record is republished on a basis of 5 minutes (but probably will make this lower). `LocalMarketMap` also has a TTL within its value that lazily removes the local mapping on demand as described in the storing and getting supplier info sequence.

#### Bootstrapping
Bootstrapping [on an interval basis is recommended](https://github.com/libp2p/rust-libp2p/issues/4730). We bootstrap on a 10 minute interval basis, but we can always change this (maybe allow lib user to change it with Config?)

We also allow a library user to accept a set of boot nodes to bootstrap into at the start. The bootnodes will typically be some hardcoded group of central servers that we can use to bootstrap, but we currently have no owned servers so users can provide their own bootnodes.

#### Encrypted Communication
The communication is encrypted with using the `noise` protocol that is currently unstable. This is subject to change to use TLS since we believe that the peer node team is using TLS(?) which we'll replace the noise protocol with and perhaps allow the user to configure that.

#### Storing and getting supplier info
How we get and store supplier info without worrying (yet) about consensus of the market (but will look into how Filecoin (might) do this if it is somewhat of an equivalent design) is through the use of provider records and the `request_response` protocol.
1. Every peer contains what we call a `LocalMarketMap` which simply stores a filehash to a tuple that contains supplier info and the creation time.  Note that this is supplier info specific to the peer itself that we're storing, and since there's typically just one user to market server to exactly one peer node in `market_dht`, we could have tightly coupled the supplier info to the peer itself (but we don't), and there would be no need for a `LocalMarketMap`. The `LocalMarketMap` has no ties to the DHT provider records itself, but the purpose of it is for the peer itself since provider records don't include itself as a provider in its list. This is called every time the `check_holders` method is called and it lazily cleans up the `market_map`, meaning that it only removes it locally on demand.
2. So, on a call to the `register_file` method, where a user/peer provides a filehash and its corresponding supplier info, this sends a command through a `tokio` mpsc bridge channel in which the `Coordinator` will receive. This typically goes for any commands sent from the `user/peer` client to the `Coordinator`, in which the `Coordinator` receives in a `select!` loop. Now, the data stored over the channel provides request information, which, in this case, includes a filehash and supplier info that gets stored in two areas: the LocalMarketMap and the DHT itself through a provider record. We only provide the filehash to the DHT, which calls the `add_provider(filehash)` method that is part of the DHT. This now stores a ProviderRecord inside the DHT, which would then send the provider record to peers close to the key.
3. Now, for a call to `check_holders`, this requires a filehash from the peer client that is received by the `Coordinator`. This, then, calls the `get_providers(filehash)` method which will return a list of provider peers that provide the key (or specifically , in our case, ones that provide the supplier info for a file). See limitations section for more on this. Now, with the provider peers returned which is just a list of provider peer IDs, we make use of the `request_response` protocol where peers can send a request to another peer; this request will be a peer ID and the response will be the supplier info that the peer provides by pulling information from its `LocalMarketMap` and sending that as the response to the requester peer.

#### Limitations
However, a key limitation to the above is that the `get_providers` method cannot return the entire market as it's limitation is based on the `K_VALUE` in the `k_buckets`. And `get_providers` only gets some subset of the peers that are close to such a key, assuming a provider exists for the filehash.

Some alternatives to this is that
1. Perhaps another method for the peer where they can directly send a request to a peerId that may or may not be a provider. The important case is if they have it; then, one approach could be to send a request to that peerId and receive the response for the supplier info for just that specific peerId. But, this would require that someone still keeps track of the entire market. This would also require that a peer must dial or possibly another bootstrap call to that peerId if it currently does not connect with that peer.
2. Implement and integrate some kind of consensus protocol for the entire market server. Maybe look into how Filecoin does it? Integrate something like Raft?
   
### `market_server`
Provides somewhat of an example for using the `market_dht::peer::Peer` struct. We allow the Peer to be constructed in sync or async contexts (which could be useful?). The market service takes ownership of this field and further uses the provided methods in an async context inside the defined gRPC methods.

From the repository's root, you can this `market_server` started with
```
cargo run --bin market_server
```
This, by default, runs the `market_server` to listen for gRPC communications on 8080 and uses 16899 for the port that the peer is listening to. You can pass in the `--help` for more configuration arguments, like so:
```
cargo run --bin market_server -- --help
```
Of course, this only runs in `debug` mode and you can build it with passing the `--release` flag (TODO: make a release profile)

### `market_test_client`
Simply, just a test client for communicating with the gRPC. It requires the user to pass in flags at the start for the supplier info. It, also, requires that the user pass in what the market gRPC port as it assumes that the market is running locally since this is its own peer (maybe change in the future?). You can get started with
```
cargo run --bin market_test_client -- --username <USERNAME> --price <PRICE> --client-port <CLIENT_PORT> --client-ip <CLIENT_IP> --market-port <MARKET_PORT>
```
Note, the market_port defaults to 8080 if nothing is provided.
#### Limitations
- Nothing prevents two clients from connecting to the same market server. This means that two different clients will be creating files under the same market_service and, consequently, under the same `peer_id`.

## TODOs
Other than the TODOs mentioned in the section above, some TODOs here

- [ ] Implement some form of CI for the tests
- [ ] Write integration tests for `market_dht`; should be simple since peers can use their own dedicated thread
- [ ] Improve the READMEs for each crate
- [ ] Look into remote connections and potentially using things like AutoNAT, maybe upnp?, etc. for dealing with peers that are protected from a NAT or firewall.
- [ ] Maybe make this just one crate specific to being a library and have any gRPC server implementation or even just embed this library inside a peer client without the gRPC overhead.
- [ ] Not necessarily on our end, but hopefully also get some centralized repository for protobuf files rather than everyone having to create their own copy of it? Then, we can easily use submodules and make PRs to those files if change is needed.
- [ ] Add more documentation inside the `market_dht` crate
- [ ] Look into the limitations above
- [ ] Fix bootnode generics
- [ ] Test with using as a docker container
- [ ] Use tracing instrumentation on logged areas
- [ ] Update connection idle timeout to prob longer

### CODE REVIEW

File: market_test_client/Cargo.toml
Comment (daminals) [inline market_test_client/Cargo.toml:1]: Perhaps a preferred way to test would be using the orcanet-rust peer node client instead, to make sure the latest versions of the market and peer node will always be interoperable

**Review by xiej2520 — COMMENTED** [review]
Rename code_doc.txt to a markdown file so formatting works.

File: market_server/src/cli.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+use clap::Parser;
+
+use crate::Port;
+
+#[derive(Debug, Parser)]
+#[command(author, version, about, long_about = None)]
+pub struct Cli {
+    #[arg(short, long, default_value = "8080")]
```
Comment (daminals) [inline market_server/src/cli.rs]: Default port for the market was previously established to be: 50051
Should modify this for consistency

**Review by daminals — APPROVED** [review]
Was able to get this running locally and interoperate with the producer and consumer nodes. Perhaps a market team can offer more focused feedback in regards to the KAD implementation, but LGTM

File: market_test_client/Cargo.toml
Comment (reaovyd) [reply market_test_client/Cargo.toml:1]: Got it will do. Thanks! Just had that `market_test_client` for sanity checking

**Comment by daminals** [comment]
Referring to a previous conversation about interoperability between the peer node and the market, in this pr the id parameter in the proto file is now managed entirely by the market, assigning a producer an id based solely off the values of  (username, ip, port, price). The proto file should be updated to remove the id parameter since it is no longer needed.

**Comment by reaovyd** [comment]
> Referring to a previous conversation about interoperability between the peer node and the market, in this pr the id parameter in the proto file is now managed entirely by the market, assigning a producer an id based solely off the values of (username, ip, port, price). The proto file should be updated to remove the id parameter since it is no longer needed.

Will do. We currently use submoduling for this into where the actual original protofile lies (within the orcanet-market-go repo), so would need some kind change for that (perhaps a PR for that file or we'll just fork off of that and submodule that repo until we get a central repo for protofiles).

**Comment by xiej2520** [comment]
The market test client should have its default market port set to 50051 to match the current market server default.

I'm pretty bootstrap doesn't need to be manually called, there's an option to set a bootstrap interval and it does so automatically.

Otherwise, the code looks good to me.

**Comment by reaovyd** [comment]
> The market test client should have its default market port set to 50051 to match the current market server default.
> 
> I'm pretty bootstrap doesn't need to be manually called, there's an option to set a bootstrap interval and it does so automatically.
> 
> Otherwise, the code looks good to me.

Thanks @xiej2520. Will change the default market port in the client to 50051. 

Correct me if I'm wrong here, but I think the auto refresh bootstrap is currently available in the master branch of the libp2p crate, but it isn't available in the [current latest release on crates.io](https://crates.io/crates/libp2p) which we're using.

**Pull request closed without merging**